statefun-sdk-js/src/core.ts

```diff
@@ -0,0 +1,329 @@
+    static fromTypeNameId(typename: string, id: string) {
+        if (isEmptyOrNull(id)) {
+            throw new Error("id must be a defined string");
+        }
+        const {namespace, name} = parseTypeName(typename);
+        return new Address(namespace, name, id, typename)
+    }
+
+
+    /**
+     * @returns {string} returns the type name string (typename) of this function.
+     */
+    get typename() {
+        return this.#typename;
+    }
+
+    /**
+     * @returns {string} the id part of the address.
+     */
+    get id() {
+        return this.#id;
+    }
+
+    /**
+     *
+     * @returns {string} returns the namespace part of this address
+     */
+    get namespace() {
+        return this.#namespace;
+    }
+
+    /**
+     *
+     * @returns {string} returns the typename's name part of this address.
+     */
+    get name() {
+        return this.#name;
+    }
+}
+
+
+/**
+ * A representation of a single state value specification.
+ * This is created from the following object:
+ * {
+ *     name: string,
+ *     type: Type,
+ *     expireAfterCall / expireAfterWrite : int
+ * }
+ */
+export interface ValueSpecOpts {
+    name: string;
+    type: Type<any>;
+    expireAfterCall?: number;
+    expireAfterWrite?: number
+}
+
+/**
+ * A representation of a single function.
+ * This can be created with the following object:
+ * {
+ *     typename: "foo.bar/baz",
+ *     fn(context, message) {
+ *         ...
+ *     },
+ *     specs: [..]
+ * }
+ */
+export interface FunctionOpts {
+    typename: string;
+    fn: (context: Context, message: Message) => void | Promise<void>;
+    specs?: ValueSpecOpts[]
+}
+
+/**
+ * an internal representation of function spec
+ */
+class ValueSpec implements ValueSpecOpts {
+    readonly #name: string;
+    readonly #type: Type<any>;
+    readonly #expireAfterCall: number;
+    readonly #expireAfterWrite: number;
+
+    constructor(name: string, type: Type<any>, expireAfterCall?: number, expireAfterWrite?: number) {
+        this.#name = name;
+        this.#type = type;
+        this.#expireAfterCall = expireAfterCall || -1;
+        this.#expireAfterWrite = expireAfterWrite || -1;
+    }
+
+    /**
+     * Creates a ValueSpec.
+     *
+     * @param {string} name the unique state name to use. Must be lowercase a-z or _.
+     * @param {Type} type the statefun type to associated with this state.
+     * @param {int} expireAfterCall the time-to-live (milliseconds) of this value after a call
+     * @param {int} expireAfterWrite the time-to-live (milliseconds) of this value after a write
+     * @returns {ValueSpec}
+     */
+    static fromOpts({name, type, expireAfterCall, expireAfterWrite}: ValueSpecOpts) {
+        if (isEmptyOrNull(name)) {
+            throw new Error("missing name");
+        }
+        if (!/^[_a-z]+$/.test(name)) {
+            throw new Error(`a name can only contain lower or upper case letters`);
+        }
+        if (type === undefined || type === null) {
```

Comment:
In TS this cannot happen anymore, but I guess this is worth keeping for JS users.